clang/lib/AST/ASTContext.cpp

```diff
@@ -14296,6 +14296,31 @@ void ASTContext::getFunctionFeatureMap(llvm::StringMap<bool> &FeatureMap,
   }
 }
 
+static SYCLKernelInfo BuildSYCLKernelInfo(ASTContext &Context,
+                                          CanQualType KernelNameType,
+                                          const FunctionDecl *FD) {
+  return { KernelNameType, FD };
+}
+
+void ASTContext::registerSYCLEntryPointFunction(FunctionDecl *FD) {
+  assert(!FD->isInvalidDecl());
+  assert(!FD->isDependentContext());
+
+  const auto *SKEPAttr = FD->getAttr<SYCLKernelEntryPointAttr>();
+  assert(SKEPAttr && "Missing sycl_kernel_entry_point attribute");
+
+  CanQualType KernelNameType = getCanonicalType(SKEPAttr->getKernelName());
+  auto IT = SYCLKernels.find(KernelNameType);
+  if (IT != SYCLKernels.end()) {
+    if (!declaresSameEntity(FD, IT->second.GetKernelEntryPointDecl()))
+      llvm::report_fatal_error("SYCL kernel name conflict");
```

Comment:
This absolutely should not be a fatal error here, this should be a diagnosed Sema-esque error.

Reply:
Very much agreed. Diagnostics will be available in the next PR. There are still a few issues we're working to fix before that PR is submitted though. Our WIP is staged [here](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr1...tahonermann:llvm-project:sycl-upstream-fe-pr2). Feel free to take a look and offer any comments here; I'll then address them before that PR is submitted.

Reply:
This isn't intended to be a temporary fatal error. Diagnostics should be (and are in the [staged changes for the next PR](https://github.com/tahonermann/llvm-project/compare/sycl-upstream-fe-pr1...tahonermann:llvm-project:sycl-upstream-fe-pr2)) issued before registration is attempted. This is intended to catch future bugs where diagnostics are missing. I chose `report_fatal_error()` over `assert()` to ensure such problems are revealed in release builds since a conflicting registration would indicate a bad situation that should never be permitted to escape until run-time.

Reply:
That is just not something we do in Clang, we use fatal error for situations where we SHOULD have handled it, but don't (yet).  I believe this to be the wrong choice.

Reply:
I would not expect a function prefixed with `register` to issue a diagnostic or report a fatal error, I would expect that to be handled by the caller, which is what it looks like the follow-up patch does.

I would expect this to be an assert (and we probably should write some documentation in the internals manual about this). Generally speaking, we prefer:

* issuing a user-facing diagnostic when the user has done something wrong
* using an assert when we think a situation exhibits a Clang developer holding an API incorrectly but we have a reasonable recovery path
* using a fatal error when we think a situation exhibits a Clang developer holding an API incorrectly and there's no possible way to recover from it
* using llvm_unreachable for code paths which should never be reached

This situation seems like an assert because 1) someone called the register function when they shouldn't have, the caller should have reported this as an error to the user (and that's what's happening in a follow-up), and 2) we have a reasonable recovery path, which is to not register the entrypoint.

Reply:
Aaron: yeah, that reflects my thoughts on this within some nits on that list that are unimportant to the discussion.

Reply:
This PR doesn't provide enough functionality to demonstrate what would happen if Aaron's suggested recovery path was implemented. Nor do we have enough staged changes to easily perform a full end-to-end test. However, I was able to experiment with the set of changes we currently have staged [here](https://github.com/tahonermann/llvm-project/compare/main...tahonermann:llvm-project:sycl-upstream-fe-pr4). To do so, I removed the disputed call to `llvm::report_fatal_error()` (along with another one currently present in the linked staged changes) and the diagnostics that protect against that call and compiled the following test case.
```
template<typename KN, typename KT>
[[clang::sycl_kernel_entry_point(KN)]]
void skep(KT k) { k(); }
void f() {
  skep<struct K>([]{});
  skep<struct K>([]{});
}
```

Compilation was successful (it shouldn't have been of course; it succeeded because the diagnostics were elided). I then checked what symbols were emitted for device compilation. A little tangent first. As noted in the documentation added by this PR, ODR-use of a `sycl_kernel_entry_point` attributed function triggers emission of an offload kernel entry point function. The staged changes emit that function named as though it were a specialization of a function template named `__sycl_kernel_caller` with a single type template argument corresponding to the SYCL kernel name. Going into this experiment, I expected to end up with some kind of error or failed assertion related to an attempt to generate a duplicate symbol. But instead, some kind of function versioning kicked in.
```
$ nm t2.o
...
0000000000000020 T _Z20__sycl_kernel_callerIZ1fvE1KEvv
0000000000000040 T _Z20__sycl_kernel_callerIZ1fvE1KEvv.1
...
```

I'm not in a position to test with an actual SYCL run-time library, so I can only speculate here, but this looks like a scenario that I was worried about. Two entry point functions were emitted, but the SYCL run-time would only resolve the kernel name to one of them with the result that the wrong entry point would be silently invoked for some cases (had actual SYCL kernel invocation functions been used in the example). That would not be fun to debug.

We certainly can (and will) change our staged code to avoid the function versioning that is happening here; we should detect the situation and emit some kind of duplicate definition diagnostic or similar. With that kind of backup in place, I would feel more comfortable replacing the call to `llvm::report_fatal_error()` with an assert. But this is an example of why I have felt that we should do more to prevent such Bad Things from happening.

Reply:
I would expect us to use the normal diagnostics engine on the 2nd to last line there to diagnose that.  I think we did something similar in `intel/llvm` that should be what you're looking for.  However, any such diagnostic is going to be imperfect cross-TU, but `report_fatal_error` cant do that anyway.  Perhaps there is a need for a `checkSYCLEntryPointFunction` that does this look up vs previous ones.

Reply:
Agreed. That diagnostic is implemented in the staged changes for the next PR; it is one of the ones I had to disable to produce the (surprising to me) symbol versioning result I described. Again, my goal here was to reduce the negative consequences of a missed diagnostic; I think an ICE is preferred over the possibility of an incorrect kernel being called at run-time.

I'm following up internally to do something about the implicit symbol versioning issue. We should have that possibility eliminated before the code generation PR is submitted.

> However, any such diagnostic is going to be imperfect cross-TU, but report_fatal_error cant do that anyway.

Indeed. We'll need to rely on linkers rejecting duplicate symbol definitions for that.

Reply:
I pushed changes to replace the call to `llvm::report_fatal_error()` with `assert()` following discussion with other colleagues. I still disagree with this direction. As a user, I would much prefer that the compiler fail catastrophically than continue merrily along generating a binary that doesn't behave as expected (In these days of safety consciousness, we would really rather continue into UB land than abort? Really?) . I acknowledge such eventualities will only occur if we fail to properly diagnose elsewhere, so the stakes are hopefully not so high. Still, I reserve the right to be unhappy and say, "I told you so" while vigorously wagging a finger if we end up getting a bug report detailing someone's difficult experience debugging why their application was mysteriously invoking the wrong SYCL kernel. I'm now resolving this conversation.